diff --git a/buildSrc/private/src/main/kotlin/org/jetbrains/androidx/build/JetBrainsMavenCoordinatesChanger.kt b/buildSrc/private/src/main/kotlin/org/jetbrains/androidx/build/JetBrainsMavenCoordinatesChanger.kt
index 7dc11446c..3340d17af 100644
--- a/buildSrc/private/src/main/kotlin/org/jetbrains/androidx/build/JetBrainsMavenCoordinatesChanger.kt
+++ b/buildSrc/private/src/main/kotlin/org/jetbrains/androidx/build/JetBrainsMavenCoordinatesChanger.kt
@@ -25,7 +25,14 @@ fun Project.changeMavenCoordinatesToJetBrains() {
     val versions = JetBrainsVersionsService.versions(project)
 
     val group = JetBrainsPublication.mavenGroupFor(path)
-    val version = Version(versions.versionOf(component.library()))
+    val version = Version(
+        if (path in setOf(":compose:ui:ui", ":compose:ui:ui-uikit")) {
+            findProperty("vlr.accessibility.version")?.toString()
+                ?: versions.versionOf(component.library())
+        } else {
+            versions.versionOf(component.library())
+        }
+    )
     this.group = group
     this.version = version
 
diff --git a/compose/ui/ui-uikit/build.gradle b/compose/ui/ui-uikit/build.gradle
index be30cfceb..7bff11d6f 100644
--- a/compose/ui/ui-uikit/build.gradle
+++ b/compose/ui/ui-uikit/build.gradle
@@ -127,7 +127,8 @@ private def configure(target, isDevice, architecture, testTarget) {
                     "SKIP_INSTALL=NO",
                     "BUILD_LIBRARY_FOR_DISTRIBUTION=YES",
                     "VALID_ARCHS=${architecture}",
-                    "MACH_O_TYPE=staticlib"
+                    "MACH_O_TYPE=staticlib",
+                    "IPHONEOS_DEPLOYMENT_TARGET=15.0"
             )
         }
 
